reps/2023-08-18-serve-java-dag-api.md

```diff
@@ -0,0 +1,266 @@
+```java
+// Demo 4
+import io.ray.serve.api.Serve;
+import io.ray.serve.deployment.Application;
+import io.ray.serve.deployment.DAGDriver;
+import io.ray.serve.deployment.InputNode;
+import io.ray.serve.generated.DeploymentLanguage;
+import io.ray.serve.handle.RayServeHandle;
+
+public class Model {
+
+  private int weight;
+
+  public Model() {}
+
+  public Model(int weight) {
+    this.weight = weight;
+  }
+
+  public int forward(int input) {
+    return weight + input;
+  }
+
+  public static void main(String[] args) throws Exception {
+    Application m1 = Serve.deployment().setDeploymentDef(Model.class.getName()).bind(1);
+    Application m2 = Serve.deployment().setDeploymentDef(Model.class.getName()).bind(2);
+
+    try (InputNode userInput = InputNode.create()) {
+      Application m1Output = m1.method("forward").bind(userInput.get(0));
+      Application m2Output = m2.method("forward").bind(userInput.get(1));
+      Application combineOutput =
+          Serve.deployment()
+              .setDeploymentDef("deployment_graph.combine")
+              .setLanguage(DeploymentLanguage.PYTHON)
+              .bind(m1Output, m2Output, userInput.get(2));
+
+      Application graph = DAGDriver.bind(combineOutput);
+      RayServeHandle handle = Serve.run(graph);
+    }
+  }
+}
+
+```
+### Cross-language transmission of DAGDriver
+In the above examples, both Demo 1 and Demo 2 are deployments of regular Java applications, which are relatively easy to implement. However, for Demo 3 and Demo 4, they involve Python `DAGDriver`, where the input of `DAGDriver` may contain `RayServeDAGHandle` that carries information for graph execution. In order to fully support graph execution orchestration through Python `DAGDriver`, it would require support for cross-language transmission of several internal core types, such as `RayServeDAGHandle`, `DeploymentMethodExecutorNode`, `DeploymentFunctionExecutorNode`, `InputAttributeNode`, and `InputNode`. This could be a significant change and needs further evaluation.
+
+### Deploying deployments of other languages through Python API
+In another REP ([Add Cpp Deployment in Ray Serve](https://github.com/ray-project/enhancements/pull/34)), it is mentioned how to deploy C++ deployments through Python. Deploying Java deployments through Python is similar. Since Java and C++ do not have the decorator mechanism like Python, a straightforward way is to directly instantiate the corresponding Deployment object:
+
+```python
+deployment_config = DeploymentConfig()
+deployment_config.deployment_language = JAVA # or CPP
```

Comment:
these are private APIs, so this won't work. the alternative below looks reasonable to me, though we should not use the private `_func_or_class` kwarg. it can just be the first positional.

Reply:
adjusted